### PR TITLE
Migrate Control Plane to codegen service client

### PR DESCRIPTION
**Pull request by guilload**
### Description
Per title.

### How was this PR tested?
- Added unit tests
- Started cp + indexer and verified that indexing plans are scheduled

### CODE REVIEW

File: quickwit/quickwit-control-plane/Cargo.toml [hunk 1]
```diff
@@ -13,34 +13,42 @@ documentation = "https://quickwit.io/docs/"
 anyhow = { workspace = true }
 async-trait = { workspace = true }
 chitchat = { workspace = true }
+dyn-clone = { workspace = true }
 itertools = { workspace = true }
+mockall = { workspace = true, optional = true }
```
Comment (fmassot) [inline quickwit/quickwit-control-plane/Cargo.toml:18]: move mockall in dev-dependencies

File: quickwit/quickwit-metastore/Cargo.toml [hunk 2]
```diff
@@ -13,6 +13,7 @@ documentation = "https://quickwit.io/docs/"
 anyhow = { workspace = true }
 async-trait = { workspace = true }
 byte-unit = { workspace = true }
+dyn-clone = { workspace = true }
```
Comment (fmassot) [inline quickwit/quickwit-metastore/Cargo.toml:16]: not sure if it's needed

**Comment by guilload** [comment]
This is ready for review.

File: quickwit/quickwit-common/src/pubsub.rs [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    #[allow(dead_code)]
+    subscription_id: usize, // Used for the `Debug` implementation.
+    subscriber: Box<dyn EventSubscriber<E>>,
+}
+
+#[derive(Debug)]
+pub struct EventSubscriptionHandle<E: Event> {
+    subscription_id: usize,
+    broker: Weak<InnerEventBroker>,
+    _phantom: PhantomData<E>,
```
Comment (fmassot) [inline quickwit/quickwit-common/src/pubsub.rs:119]: `todo!()`!

Comment (guilload) [reply quickwit/quickwit-common/src/pubsub.rs:119]: `cancel` consumes `self`, which subsequently calls `drop` on `self`.

File: quickwit/quickwit-control-plane/Cargo.toml (hunk 1 above)
Comment (guilload) [reply quickwit/quickwit-control-plane/Cargo.toml:18]: I think it's correct to add `mockall` as an optional dep if the feature `testsuite` depends on it.

File: quickwit/quickwit-common/src/type_map.rs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::any::{Any, TypeId};
+use std::collections::HashMap;
+
+#[derive(Debug, Default)]
+pub struct TypeMap(HashMap<TypeId, Box<dyn Any + Send + Sync + 'static>>);
+
+impl TypeMap {
+    pub fn contains<T: Any + Send + Sync + 'static>(&self) -> bool {
```
Comment (fmassot) [inline quickwit/quickwit-common/src/type_map.rs:27]: I think you can remove `static` as it's already insured by `Any`.

Comment (fmassot) [inline quickwit/quickwit-common/src/type_map.rs:27]: why not keeping `contains_key` name?

Comment (fmassot) [reply quickwit/quickwit-common/src/type_map.rs:27]: got it, `contains` is better.

File: quickwit/quickwit-serve/src/lib.rs [hunk 5]
```diff
@@ -114,6 +115,12 @@ fn has_node_with_metastore_service(members: &[ClusterMember]) -> bool {
 }
 
 pub async fn serve_quickwit(config: QuickwitConfig) -> anyhow::Result<()> {
+    let mut registry = Registry::for_namespace(&config.node_id);
```
Comment (fmassot) [inline quickwit/quickwit-serve/src/lib.rs:118]: I may miss something but currently we don't need this registry right? is it part of a hidden plan? :)

File: quickwit/quickwit-common/src/pubsub.rs [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+            .inner
+            .subscriptions
+            .lock()
+            .expect("The lock should never be poisoned.");
+
+        if let Some(typed_subscriptions) = subscriptions.get::<EventSubscriptions<E>>() {
+            for subscription in typed_subscriptions.values() {
+                let event = event.clone();
+                let mut subscriber = subscription.subscriber.clone();
+                tokio::spawn(tokio::time::timeout(Duration::from_secs(600), async move {
```
Comment (fmassot) [inline quickwit/quickwit-common/src/pubsub.rs:100]: don't know what would be a good timeout duration.

Nitpick: log a warning if we have a timeout
Nitpick: I would reduce the timeout but I don't have any good reasons, it's just that I would be scared if a handle_event can last 600 seconds...

File: quickwit/quickwit-control-plane/src/lib.rs [hunk 7]
```diff
@@ -17,17 +17,67 @@
...
+
+impl From<AskError<ControlPlaneError>> for ControlPlaneError {
+    fn from(error: AskError<ControlPlaneError>) -> Self {
+        match error {
+            AskError::ErrorReply(error) => error,
+            AskError::MessageNotDelivered => {
+                ControlPlaneError::Unavailable("Request not delivered".to_string())
+            }
+            AskError::ProcessMessageError => {
+                ControlPlaneError::Internal("Request failed".to_string())
```
Comment (fmassot) [inline quickwit/quickwit-control-plane/src/lib.rs:76]: nitpick: `Error while processing request`

File: quickwit/quickwit-control-plane/src/lib.rs [hunk 8]
```diff
@@ -41,3 +91,70 @@ pub async fn start_control_plane_service(
...
+/// Note: we don't need to send an event to the control plane on index creation.
+/// A new index has no source and thus will not change the scheduling of indexing tasks.
+// TODO(fmassot):
+// - Forbid a `MetastoreWithControlPlaneTriggers` that wraps a gRPC client metastore.
+// - We don't sent any data to the Control Plane. It could be nice to send the relevant data to the
+//   control plane and let it decide to schedule or not indexing tasks.
+#[async_trait]
+impl EventSubscriber<MetastoreEvent> for ControlPlaneServiceClient {
+    async fn handle_event(&mut self, event: MetastoreEvent) {
+        let event = match event {
```
Comment (fmassot) [inline quickwit/quickwit-control-plane/src/lib.rs:108]: Nitpick: you could log the metastore event on error.

**Review by fmassot — COMMENTED** [review]
LGTM. Can you look at my comments on the registry stuff?

**Review by fmassot — APPROVED** [review]

File: quickwit/quickwit-common/src/type_map.rs (hunk 4 above)
Comment (guilload) [reply quickwit/quickwit-common/src/type_map.rs:27]: Cool.

File: quickwit/quickwit-serve/src/lib.rs (hunk 5 above)
Comment (guilload) [reply quickwit/quickwit-serve/src/lib.rs:118]: Yeah, the registry is unnecessary for this PR, but I plan to use it in future works. I can add the registry later and hold the event broker in `QuickwitServices` for the time being.

File: quickwit/quickwit-common/src/registry.rs [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+
+use std::any::Any;
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+
+use once_cell::sync::Lazy;
+
+use crate::type_map::TypeMap;
+
+static REGISTRIES: Lazy<Arc<Mutex<HashMap<String, Registry>>>> = Lazy::new(Default::default);
```
[the remark this answers is not in the document]
Comment (fmassot) [reply quickwit/quickwit-common/src/registry.rs:28]: Yes, it's fine with the `for_namespace`, and yes it's not the same of for the ingest API. 

In fact the main point is that I don't see yet how you plan to use this, that's why I'm a bit reluctant on this finally.

I don't see any risk in merging the PR though.

**Pull request merged**